python/tests/ibd.py

```diff
@@ -0,0 +1,485 @@
+            s = "[{}, ..., {}]".format(repr(self.head), repr(self.tail))
+        return s
+
+    def __len__(self):
+        # Returns the number of segments in the list.
+        count = 0
+        seg = self.head
+        while seg is not None:
+            count += 1
+            seg = seg.next
+        return count
+
+    def __add__(self, other):
+        """
+        Use to append another SegmentList, or a single segment.
+        SegmentList1 + SegmentList2 will modify SegmentList1 so that
+        SegmentList1.tail.next = SegmentList2.head
+        SegmentList1 + Segment1 will add Segment1 to the tail of SegmentList1
+        NOTE: both of these operations will modify SegmentList1 in place!
+        So + functions like +=
+        """
+        assert isinstance(other, SegmentList) or isinstance(other, Segment)
+
+        if isinstance(other, SegmentList):
+            if self.head is None:
+                self.head = other.head
+                self.tail = other.tail
+            else:
+                self.tail.next = other.head
+                self.tail = other.tail
+        elif isinstance(other, Segment):
+            if self.head is None:
+                self.head = other
+                self.tail = other
+            else:
+                self.tail.next = other
+                self.tail = other
+
+
+class IbdFinder:
+    """
+    Finds all IBD relationships between specified samples in a tree sequence.
+    """
+
+    def __init__(self, ts, samples=None, min_length=0, max_time=None):
+
+        self.ts = ts
+        # Note: samples *must* be in order of ascending node ID
+        if samples is None:
+            self.samples = ts.samples()
+        else:
+            self.samples = samples
+        self.samples.sort()
+        self.min_length = min_length
+        self.max_time = max_time
+        self.current_parent = self.ts.tables.edges.parent[0]
+        self.current_time = 0
+        self.A = [SegmentList() for _ in range(ts.num_nodes)]  # Descendant segments
+        self.SegL_head = SegmentList()
+        self.SegL_tail = SegmentList()
+        self.tables = self.ts.tables
+        self.parent_list = [[] for i in range(0, self.ts.num_nodes)]
+        for e in self.ts.tables.edges:
+            if (
+                len(self.parent_list[e.child]) == 0
+                or e.parent != self.parent_list[e.child][-1]
+            ):
+                self.parent_list[e.child].append(e.parent)
+        # Objects below are needed for the IBD segment-holding object.
+        self.sample_id_map = self.get_sample_id_map()
+        self.num_samples = len(self.samples)
+        self.sample_pairs = self.get_sample_pairs()
+        """
+        Note: in the C code the object below should be a struct array.
+        Each item will be accessed using its index, which corresponds to a particular
+        sample pair. The mapping between index and sample pair is defined in the
+        find_sample_pair_index method further down.
```

Comment:
Hm, this is in `N^2`-land, where I don't think we want to be? Maybe it'll be the case that most pairs have a segment, but also maybe not, and for big datasets this will be bad. So, in python you don't need to initialize all `N^2` empty dicts (but have to check below if the pair is already in `self.ibd_segemtns`). In C you can use an AVL tree of segment lists sorted on the pair of IDs to do the same thing.

Reply:
There's just a single dict, but yes, the number of keys in the dict is n choose 2 where the user has inputted n samples. In the call with @jeromekelleher last night we discussed initialising them with `None` instead of with empty SegmentLists, which will hopefully solve some of this big memory problem?

I think it makes sense to keep all sample pairs in the outputted dict, even if there are no IBD segments, because we'll have to do that in C as well. In C the output will be stored as a struct array, where we access the list of IBD segments corresponding to a particular pair using a predefined lexicographic ordering. It will be confusing and difficult to define an ordering based only on the sample pairs that happen to have non-empty IBD output, as that will differ from run to run, and it won't be known until the algorithm has been applied in any case.

Reply:
Well, we'll see, but I don't think you want to N^2 anything, including the size of your struct array. That's why I'm suggesting the AVL tree, which we have a predefined implementation of (see `avl.h`) and lets you do log(N) insertions and lookups easily: you don't have to worry about the ordering. It'll be easy enough to switch in later, I suppose, so maybe you should go forward with the current plan.